x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/helpers.ts

```diff
@@ -389,6 +391,25 @@ export const processK8sUsernames = (clusterId: string, event: TelemetryEvent): T
   return event;
 };
 
+export const processDetectionRuleCustomizations = (event: TelemetryEvent) => {
+  const ruleSource = event['kibana.alert.rule.parameters']?.rule_source;
+  if (
+    !ruleSource ||
+    ruleSource.type === 'internal' ||
+    ruleSource.is_customized === false ||
+    ruleSource.has_base_version === false
+  ) {
+    return undefined; // Don't return anything if rule is not customized or base version doesn't exist
+  }
+  const numberOfFunctionalFields = ruleSource.customized_fields.filter(
+    (field) => FUNCTIONAL_FIELD_MAP[field.field_name as keyof RuleResponse]
+  ).length;
+  return {
+    customized_fields: ruleSource.customized_fields.map((fieldObj) => fieldObj.field_name),
+    num_functional_fields: numberOfFunctionalFields,
+  };
+};
```

Comment:
This implementation doesn't handle the case when `customized_fields` is undefined. This will happen with any alerts generated prior to the release of this feature. It may even happen with alerts generated after releasing it, if they are generated by a rule which won't be updated by the user yet. When this happens, the function will throw an exception and break telemetry.

We should be very cautions about the data we process and code defensively.

Please add explicit null/undefined checks for both `customized_fields` and `has_base_version` fields and add corresponding test cases to `x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/helpers.test.ts`.

Reply:
Good catch

Reply:
This is one of the main reasons I implemented the versioned alerts schema: https://github.com/elastic/kibana/blob/main/x-pack/solutions/security/plugins/security_solution/common/api/detection_engine/model/alerts/schema.ts - it improves type safety in handling cases where we add new fields to alerts. Unfortunately, the rule schemas are not versioned so the alert schema doesn't get versioned fields for data copied from the rule. If we envision more scenarios like this in the future it could be worth investing in creating versioned rule schemas that can integrate into the alert schemas.